pkg/controller/azvolume.go

```diff
@@ -250,68 +245,59 @@ func (r *ReconcileAzVolume) triggerDelete(ctx context.Context, azVolume *azdiskv
 		if volumeDeleteRequested {
 			w.Logger().V(5).Info("Deleting Volume...")
 		}
+	}
 
-		waitCh := make(chan goSignal)
-		//nolint:contextcheck // call is asynchronous; context is not inherited by design
-		go func() {
-			_, goWorkflow := workflow.New(ctx)
-			var deleteErr error
-			defer func() { goWorkflow.Finish(deleteErr) }()
-			waitCh <- goSignal{}
-
-			goCtx := goWorkflow.SaveToContext(context.Background())
-
-			deleteCtx, deleteCancel := context.WithTimeout(goCtx, cloudTimeout)
-			defer deleteCancel()
+	waitCh := make(chan goSignal)
+	//nolint:contextcheck // call is asynchronous; context is not inherited by design
+	go func() {
+		_, goWorkflow := workflow.New(ctx)
 
-			reportError := func(obj interface{}, err error) error {
-				azv := obj.(*azdiskv1beta2.AzVolume)
-				_ = r.updateError(azv, err)
-				_, derr := r.updateState(azv, azdiskv1beta2.VolumeDeletionFailed, forceUpdate)
-				return derr
-			}
+		var deleteErr error
+		defer func() { goWorkflow.Finish(deleteErr) }()
+		waitCh <- goSignal{}
 
-			var updateFunc azureutils.UpdateCRIFunc
-			var err error
-			updateMode := azureutils.UpdateCRIStatus
+		goCtx := goWorkflow.SaveToContext(context.Background())
 
-			// Delete all AzVolumeAttachment objects bound to the deleted AzVolume
-			var attachments []azdiskv1beta2.AzVolumeAttachment
-			attachments, err = r.cleanUpAzVolumeAttachmentByVolume(deleteCtx, azVolume.Name, azvolume, azureutils.AllRoles, mode)
-			if err != nil {
-				updateFunc = func(obj client.Object) error {
-					return reportError(obj, err)
+		var updateFunc azureutils.UpdateCRIFunc
+		var err error
+		updateMode := azureutils.UpdateCRI
+
+		deleteCtx, deleteCancel := context.WithTimeout(goCtx, cloudTimeout)
+		defer deleteCancel()
+
+		// Delete all AzVolumeAttachment objects bound to the deleted AzVolume
+		var attachments []azdiskv1beta2.AzVolumeAttachment
+		attachments, err = r.cleanUpAzVolumeAttachmentByVolume(deleteCtx, azVolume.Name, azvolume, azureutils.AllRoles, mode)
+		if err == nil {
+			var wg sync.WaitGroup
+			errors := make([]error, len(attachments))
+			numErrors := uint32(0)
```

Comment:
I think it would be better to communicate with channels here. `errors` itself could be a buffered channel. We would then send all the errors thorough that channel and instead of wg.Wait () we can have 
```
go func() {
		wg.Wait()
		close(errors )
	}()
```

then we will loop over errors to check for error: 
```
for error := range errors {
		// update status
	}
```

Reply:
Yeah, I see the benefit of using a buffered channel here. Let's merge this one and I will iterate on this in a separate PR as Huichan's change has dependency on this change.

Reply:
Sure!